### PR TITLE
Add top level request method

### PR DESCRIPTION
**Pull request by franekmagiera**
My proposition for #1993 
Was not sure if this feature needed tests as it reuses already tested `PoolManager`.
Let me know what you think.

### CODE REVIEW

**Comment by shazow** [comment]
We recently had a discussion about this in our community Discord channel, here's the relevant snippet:
image

**Comment by franekmagiera** [comment]
Thanks for your response. So, should I just change `_manager` to `_DEFAULT_POOL` and it would be good to go (assuming you are ok with the code)? Or are you still considering adding other features for this?

**Comment by shazow** [comment]
I'll let @SethMichaelLarson make the final call here.

Personally I'd say let's make it `DEFAULT_POOL`, change the `urllib3.request(...)` method to be a stand-alone function that calls `DEFAULT_POOL.request(...)` while removing the `**urllib3_kw` args from the call signature.

```python
def request(...):
    return DEFAULT_POOL.request(...)
```

Also IMO, the docstring should explain that `urllib3.request` uses a module-global `DEFAULT_POOL` instance and that side effects could be shared across dependencies that rely on it. To avoid side effects, create a new PoolManager and use that instead.

I think more details about what the issue might be is better than just blankly telling people not to use it in production. :)

**Comment by pquentin** [comment]
+1 to what @shazow said.

Also, we do need a test in `test/with_dummyserver/test_poolmanager.py` to make sure nobody removes this by accident. Untested, but it could look like this:

```
    def test_top_level_request(self):
        with PoolManager() as http:
            r = http.request("GET", f"{self.base_url}/")
            assert r.status == 200
            assert r.data == b"Dummy server!"
```

**Comment by franekmagiera** [comment]
Thanks for the pointers, they have been really helpful, I added suggested changes.

I've got a check failing for Ubuntu 3.1-dev test-3.10 in `test_http.py`. for now I don't know what's causing the 'Address already in use error' - tornado's `bind_socket` is called with `None` as far as I checked and should allocate ports automatically.

Also Ubuntu 3.8 seems to be stuck on set up for some reason.

**Comment by pquentin** [comment]
Closing/reopening to see what failures are real (they don't look related to your pull request).

File: docs/user-guide.rst
Comment (sethmlarson) [inline docs/user-guide.rst:62]: We can remove the note about not being able to use in production. It's production capable given you're okay with the above constraints.

File: src/urllib3/__init__.py
Comment (sethmlarson) [inline src/urllib3/__init__.py:99]: Shouldn't we omit `**urlopen_kw` in the signature if we're going to drop them on the floor anyways?

File: src/urllib3/__init__.py
Comment (sethmlarson) [inline src/urllib3/__init__.py:92]: I'm not thrilled about using `certifi` as a default given us trying to move away from it and to instead rely on system certificates. We use system certificates by default on Linux and kinda on Windows and not on macOS iirc. Not certain what to do about this since whatever we choose we're supporting for all of 2.x.

**Review by sethmlarson — CHANGES_REQUESTED** [review]
Thanks for this change, I have some comments and questions.

**Comment by sethmlarson** [comment]
Also `DEFAULT_POOL` could either be a **Pool**Manager or Connection**Pool** when looking at the name. Also PoolManager isn't exactly "configurable" past initialization, it's making me think we actually want it to be private? @shazow @pquentin?

File: docs/user-guide.rst
Comment (shazow) [reply docs/user-guide.rst:62]: At the very least, if people find scenarios where this breaks "in production" in unexpected ways then I'd like to consider it a bug.

**Comment by shazow** [comment]
@sethmlarson I don't feel very strongly about it. It's not like the name is a lie, it is indeed a pool! The kind of pool that we choose by default is an implementation/typing detail.

I'm onboard with starting more conservative though, `_DEFAULT_POOL` is fine with me.

**Comment by pquentin** [comment]
There are good arguments either way.

Regarding CI, the tests actually passed (https://github.com/urllib3/urllib3/actions/runs/538104753) but GitHub is telling me that all checks are still expected on this PR. 🤦

File: docs/user-guide.rst
Comment (franekmagiera) [reply docs/user-guide.rst]: Removed the recommendation to not use it in production.

File: src/urllib3/__init__.py
Comment (franekmagiera) [reply src/urllib3/__init__.py]: Right, makes more sense, it is better to get an error when passing keyword arguments instead of having them silently dropped. Removed the ``**urlopen_kw`` from the signature.

File: src/urllib3/__init__.py
Comment (franekmagiera) [reply src/urllib3/__init__.py]: Initially I used certifi because of the #1993 description, but I think it would be better to use system certificates by default. Removed certifi.

**Comment by franekmagiera** [comment]
Thanks for the responses. Personally, I feel `_DEFAULT_POOL` is a natural choice for this PoolManager instance, although maybe not entirely reflecting what object it really is.

I made the suggested changes.

Also, would it make sense to have some kind of reload method for the `_DEFAULT_POOL`? Given that there is no access to this instance maybe having a method that would call ``__exit__`` and reinitialized the `_DEFAULT_POOL` could come in handy?

**Comment by shazow** [comment]
Looking good to me!

>  Given that there is no access to this instance...

Does recent Python actually enforce limiting access to underscore-prefixed namespaces? I haven't Python'd in a while but back in the day it was merely a convention, we still had full access to everything.

In either case, IMO let's err on the side of a smaller API surface until it's clear that we need more.

File: docs/user-guide.rst
Comment (franekmagiera) [reply docs/user-guide.rst]: Also added a note about system CA certificates being loaded by default.

**Comment by franekmagiera** [comment]
Right, my mistake. It will only be omitted in case of `import *`.

**Review by sethmlarson — APPROVED** [review]
Did some minor changes manually, thanks for this change. Will merge once green!

**Comment by pquentin** [comment]
@franekmagiera Thank you! Would you like to continue working on urllib3? Are there areas you're interested in specifically?

**Comment by shazow** [comment]
@franekmagiera Also thank you from me! This has been something I felt urllib3 needed for a long time, I really appreciate you diving in and making it happen. :)

**Comment by franekmagiera** [comment]
Thanks for support from your side ;)

Sure, I would be happy to tackle another issue. There is nothing specific I'm interested in. I guess most things related to v2 would be ok. Do you have any suggestions?

**Comment by sethmlarson** [comment]
We have all the issues for v2 in [this milestone](https://github.com/urllib3/urllib3/milestone/6), if you're interested in type hints then working on https://github.com/urllib3/urllib3/issues/1897 incrementally by adding types to a single module would be a good issue.

**Pull request merged**
